### PR TITLE
[Notifier] Added telnyx notifier bridge

### PR DESCRIPTION
**Pull request by krasilnikovs**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

Added telnyx notifier bridge for sending sms

https://developers.telnyx.com/docs/api/v2/messaging/Messages?lang=curl#createMessage

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (jderusse) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: 5.3

**Review by jderusse — COMMENTED** [review]
Thank you for this contribution.

Don't forget to add an entry in the ChangeLog. and in UnsupportedSchemeException files

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: Thanks, changed to 5.3

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (jschaedl) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]:
```suggestion
===============
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (jschaedl) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: Using `sprintf` would be more readable imho.

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: In other transports uses `sprintf` and for consistency I decided to use `sprintf` as well.

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (jschaedl) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: I didn't see this in my first review, but I would use sprintf here as well.

**Review by jschaedl — COMMENTED** [review]
@krasilnikovm Thank you for your adjustments. Code looks good to me now, except the one line where `sprintf` could be used as well.

I know some of the bridges got merged without tests, but I think having tests here would be really great. Do you think you could add some?

**Comment by krasilnikovs** [comment]
> 
> 
> @krasilnikovm Thank you for your adjustments. Code looks good to me now, except the one line where `sprintf` could be used as well.
> 
> I know some of the bridges got merged without tests, but I think having tests here would be really great. Do you think you could add some?

Hi @jschaedl, sorry for I'm not checked other places for using sprintf , and now I'm working on writing tests but I have noticed that SmsMessage::getOptions method always returns null, https://github.com/symfony/symfony/blob/0137609eaf04935a7762952ad6c8f26a87e32786/src/Symfony/Component/Notifier/Message/SmsMessage.php#L98-L101 
https://github.com/symfony/symfony/blob/0137609eaf04935a7762952ad6c8f26a87e32786/src/Symfony/Component/Notifier/Message/SmsMessage.php#L28-L37
 Can I add initializing options in the PR or better to create separate PR?

---------------------------------------------------------------------
Updated
I have added initializing options in the PR

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: Done and tests added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (stof) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: @fabpot should `__toString` really expose the API key ?

Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: btw, this does not look like it produces a valid DSN for the corresponding factory.

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (stof) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: I would use `4.4` as min version rather than `4.3` as 4.3 is already EOL

Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: Fixed

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: My bad, Fixed

File: src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:4]: Missing `/.gitattributes export-ignore`

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Can you add a note about the DSN format (like done by most bridges)?

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: Can be removed

File: src/Symfony/Component/Notifier/Message/SmsMessage.php
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]: This change should be removed and discussed in another PR.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by fabpot — CHANGES_REQUESTED** [review]
Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: Wouldn't it work with notifier 5.1 too?

Not sure how we handle this in general? 

cc @jschaedl @fabpot

Comment (derrabus) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: If this PR is merged, it will be part of the Symfony 5.3 release. Symfony 5.1 will be EOL by then. I wouldn't change this line.

Comment (fabpot) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: In any case, the code is not compatible with 5.1.

File: src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:4]:
```suggestion
/.gitattributes export-ignore
/.gitignore export-ignore
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]:
```suggestion
Provides [Telnyx](https://telnyx.com) integration for Symfony Notifier.
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php]:
```suggestion
        $transport = $this->createTelnyxTransport();
        $chatMessage = new ChatMessage('Hello');
        
        $this->expectException(LogicException::class);        
        
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php]: please remove `:void` from all test classes

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]:
```suggestion
        $from = $dsn->getRequiredOption('from');        
```

It looks like `from` is required

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: Please use newly introduced `UnsupportedMessageTypeException` here.

Fore reference: https://github.com/symfony/symfony/pull/39365

Thanks

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:29]: Please pull up the scheme check like here, to first check if the scheme is supported and afterwards for missing options. Thanks

https://github.com/symfony/symfony/blob/7d846d3e82f4e3dabec7d4b5300f50a1dd9ed7d6/src/Symfony/Component/Notifier/Bridge/Sinch/SinchTransportFactory.php#L27-L47

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: should use `~5.3.0`

**Comment by OskarStark** [comment]
Hi @krasilnikovm  👋 

Are you interested in finishing this PR or am I allowed to take over?

Cheers Oskar

**Comment by krasilnikovs** [comment]
> 
> 
> Hi @krasilnikovm 👋
> 
> Are you interested in finishing this PR or am I allowed to take over?
> 
> Cheers Oskar

Hi, last month I was sick, now I feel nice and I'm going to finish the pr on weekend.

**Comment by OskarStark** [comment]
> Hi, last month I was sick, now I feel nice and I'm going to finish the pr on weekend.

Glad to hear that  👍 😃

File: src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:4]: Added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: Removed

File: src/Symfony/Component/Notifier/Message/SmsMessage.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Message/SmsMessage.php]: PR created: https://github.com/symfony/symfony/pull/39365

File: src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:4]: Added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: Removed

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:29]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: Done

**Comment by krasilnikovs** [comment]
> 
> 
> Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

@fabpot Actually, I updated the file, or Am I something missed?
https://github.com/symfony/symfony/blob/1bbb34b12e5d484ce6cde5165a7ef978a38adc29/src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php#L89-L95

File: src/Symfony/Component/Notifier/Bridge/Telnyx/LICENSE
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/LICENSE]:
```suggestion
Copyright (c) 2021 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]:
```suggestion
```
TELNYX_DSN=telnyx://API_KEY@default?from=FROM
```

where:
 - `API_KEY` is your telnyx api key
 - `FROM` is your telnyx phone number
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]:
```suggestion
        $apiKey = $this->getUser($dsn);
        $from = $dsn->getRequiredOption('from');

```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: You should add yourself as an author here

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]:
```suggestion
        "psr-4": { "Symfony\\Component\\Notifier\\Bridge\\Telnyx\\": "" },
```

**Review by OskarStark — APPROVED** [review]
Ready to merge after my comments 👍

File: src/Symfony/Component/Notifier/Bridge/Telnyx/LICENSE
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/LICENSE]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php]: done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: Added

File: src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json]: Done

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/Telnyx/CHANGELOG.md
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/CHANGELOG.md]:
```suggestion
5.3
---

 * Add the bridge
```

**Review by OskarStark — CHANGES_REQUESTED** [review]
This PR is blocked until https://github.com/symfony/symfony/pull/39774 is merged!

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]:
```suggestion
```

**Comment by OskarStark** [comment]
Can you please rebase to get rid of the conflicts? Thanks

File: src/Symfony/Component/Notifier/Bridge/Telnyx/CHANGELOG.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/CHANGELOG.md]: Done

File: src/Symfony/Component/Notifier/Bridge/Telnyx/README.md
Comment (krasilnikovs) [reply src/Symfony/Component/Notifier/Bridge/Telnyx/README.md]: Done

**Comment by OskarStark** [comment]
@krasilnikovm I need to close this PR, because it uses your other PR [[Notifier] Added ability to initialize sms options](https://github.com/symfony/symfony/pull/39774) which was closed ❌  without merging. 😢 

Feel free to reopen and update this PR or reopen a new one if you found another solution without the need of `MessageOptions` for SMS.

Thanks for your patience 🚀

**Pull request closed without merging**
